src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1315 @@
+import datetime
+import enum
+import errno
+import grp
+import ipaddr
+import ldap
+import ldap.sasl
+import ntplib
+import pwd
+import socket
+import subprocess
+
+from dns import resolver
+from ldap.controls import SimplePagedResultsControl
+from middlewared.schema import accepts, Bool, Dict, Int, List, Str
+from middlewared.service import job, private, ConfigService, ValidationError, ValidationErrors
+from middlewared.service_exception import CallError
+from middlewared.utils import run
+
+
+class DSStatus(enum.Enum):
+    """
+    Following items are used for cache entries indicating the status of the
+    Directory Service.
+    :FAULTED: Directory Service is enabled, but not HEALTHY.
+    :LEAVING: Directory Service is in process of stopping.
+    :JOINING: Directory Service is in process of starting.
+    :HEALTHY: Directory Service is enabled, and last status check has passed.
+    There is no "DISABLED" DSStatus because this is controlled by the "enable" checkbox.
+    This is a design decision to avoid conflict between the checkbox and the cache entry.
+    """
+    FAULTED = 1
+    LEAVING = 2
+    JOINING = 3
+    HEALTHY = 4
+
+
+class neterr(enum.Enum):
+    JOINED = 1
+    NOTJOINED = 2
+    FAULT = 3
+
+
+class SRV(enum.Enum):
+    DOMAINCONTROLLER = '_ldap._tcp.dc._msdcs.'
+    FORESTGLOBALCATALOG = '_ldap._tcp.gc._msdcs.'
+    GLOBALCATALOG = '_gc._tcp.'
+    KERBEROS = '_kerberos._tcp.'
+    KERBEROSDOMAINCONTROLLER = '_kerberos._tcp.dc._msdcs.'
+    KPASSWD = '_kpasswd._tcp.'
+    LDAP = '_ldap._tcp.'
+    PDC = '_ldap._tcp.pdc._msdcs.'
+
+
+class SSL(enum.Enum):
+    NOSSL = 'off'
+    USESSL = 'on'
+    USETLS = 'start_tls'
+
+
+class ActiveDirectory_DNS(object):
+    def __init__(self, **kwargs):
+        super(ActiveDirectory_DNS, self).__init__()
+        self.ad = kwargs.get('conf')
+        self.logger = kwargs.get('logger')
+        return
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        return
```

Comment:
I am confused why we are making ActiveDirectory_DNS a context manager. Any future plans?

Reply:
Stopped making it a context manager.